### ISSUE TITLE
Watched.clearWhenTriggered should clear screen before first run

### ISSUE THREAD

**Issue by japgolly** [sbt/sbt#4445]
- [x] I've read the [CONTRIBUTING](https://github.com/sbt/sbt/blob/develop/CONTRIBUTING.md) guidelines

## steps

With `triggeredMessage := Watched.clearWhenTriggered` run
`~compile`

## problem

SBT doesn't clear the screen before the first run.

## expectation

SBT should doesn't clear the screen before the first run. It used to.

## notes

sbt version: 1.2.6

**Comment by eatkins** [comment]
Hmm. I didn't realize that this was the previous behavior. I'm not 100% sure that I agree that sbt was doing the right thing before. I guess it depends on the semantics of the word "triggered." To me, that means that the screen should clear when a build is triggered. One could argue that executing the command `~compile` triggers the build, but that's not how I think about it.

It's a trivial change to restore the old behavior in 1.3.x, but given that there has never really been a specification for the watch subsystem, I'd like to solicit additional feedback on what the correct behavior should be here.

**Comment by japgolly** [comment]
Cool, additional feedback is fine but to be clear, the behaviour we're discussing will only affect people who use watching and who've taken the time to set `triggeredMessage := Watched.clearWhenTriggered` so a distinction should be made in whose feedback is considered. People who don't use this feature shouldn't affect the feature for those who do. So for me personally as a user, I don't care about the names of settings or potential interpretations of semantics, what I care about is the larger SBT experience and my workflow. I've been a long time user of this feature and it's very important to my workflow.

I think it's fair to say anyone who takes the time to set `triggeredMessage := Watched.clearWhenTriggered` has the intent that when they watch a task, they want to see only the results of that task and nothing else. I don't think anyone would want to say to "well the first run is a special case and I want to see all the output that was already on the screen but from the second onwards, I only want to see the results of that task and nothing else". :smiley:

**Comment by eatkins** [comment]
I hear you, but I think you are overgeneralizing. I am an anyone who does consider the first run a special case. That may be because I tend to be a literal thinker and, as I said previously, the name of the key is triggeredMessage and, to me, a build is not triggered just by entering watch mode. I draw a clear distinction in my head between a manual build, e.g. running compile manually, and a triggered build, i.e. a build that is triggered by an external action. The definition of the task key is (and has always been):
```scala
val triggeredMessage = settingKey[WatchState => String]("The message to show before triggered execution executes an action after sources change.").withRank(DSetting)
```
 which is also more consistent with my interpretation.

I don't want to break your, or any other user's, workflow so I want to figure out how to reconcile your needs with the current implementation, which more or less literally follows the specification as described in the key (unlike the previous behavior). My objection to just reverting back to the old behavior is that I don't want to keep adding special case config options and I think both interpretations are valid.

In 1.3.0 there is a new build key:
```scala
val watchTriggeredMessage = settingKey[(TypedPath, Int) => Option[String]]("The message to show before triggered execution executes an action after sources change. The parameters are the path that triggered the build and the current watch iteration count.").withRank(DSetting)
```
What I can probably do is invoke that function with the path set to the working directory and with the build count set to zero. This would give you the control to preserve your old workflow.

Also, as you can see, printing a message before the first run is a special case, at least from a technical perspective, because there is initially no path that triggered the build.

**Comment by eatkins** [comment]
In the meantime, you can add this to your build to recreate the previous behavior:
```scala
Global / onLoad := { state =>
  state.definedCommands.find(_.nameOption.contains("~")) match {
    case Some(cmd: Command) =>
      state.copy(
        definedCommands = state.definedCommands :+ Command.make("~", cmd.help(state)) { s =>
          println(Watched.clearScreen)
          cmd.parser(s)
        })
    case _ => state
  }
}
```

**Comment by japgolly** [comment]
I think we have a lot of common ground here, especially in that we agree that the definition is ambiguous and both interpretations are valid 😀 Even now my interpretation of triggering simply means immediately scheduling execution which is what happens when you type `~compile`, it does immediately trigger an execution. But I get what you mean too and the description text of the task certainly describes that.

But yeah, the generalisation that accepts a counter I think is perfect and will make everyone happy. :+1:

And thank you very much for the immediate workaround, I'll try that when I get home.

**Comment by dwijnand** [comment]
According to https://github.com/sbt/sbt/pull/4512#issuecomment-453777507 fixed in #4512.
